### DISCUSSION TITLE
rs232 with the dumbest terminal possible

### DISCUSSION THREAD

**Discussion by Whammo** [jkotlinski/durexforth#427] [Show and tell]
I'd heard rumors of this working, but until tonight, not seen it with my own eyes.
It's quite the little terminal. At least in Vice.
Now, even faster.
Faster yet, now with GETIN on both channels.
Slimmed up, checking ST
 Resolved Race condition.
Removed unnecessary clrchn's 
Real terminal program has bell.
~Can't get it reliable on latest release of Vice emu.~
 Works at 1200 baud, drops characters.

```forth
require io
code getin ( -- chr )
dex, w stx, 0 lda,# msb sta,x
$ffe4 jsr, \ GETIN
w ldx, lsb sta,x
;code
code bell
$15    lda,#
$d418  sta,
$09    ldy,#
0      ldx,#
$d405  sty,
$d406  stx,
$30    lda,#
$d401  sta,
$20    lda,#
$d404  sta,
$21    lda,#
$d404  sta,
;code

: st $0297 c@ ;
$ff99 constant memtop
0 sr c! latest $100 - xr ! memtop sys
\ open 2400
: dt page 2 device $0a here c!
here 1 16 0 open 2drop
\ read keys and transmit
begin getin ?dup if 16 chkout
ioabort emit clrchn then
16 chkin ioabort
\ receive and print
begin getin ?dup while
dup 7 = if drop bell
else emit then repeat
clrchn \ drop silently
\ check stop key, errors
$91 c@ $7f = st $50 and or until
16 close 8 device quit ;


```
As fast as I can:
```forth
code dt
w      stx, latest dup $ff and
       ldx,#
100/   ldy,#
       clc,
$ff99  jsr, \ memtop
$93    lda,#
$ffd2  jsr, \ page
$0a    lda,# latest $100 - dup dup
       sta,    ( 2400 baud )
0      lda,# 1+
       sta,
2      lda,# dup $ff and
       ldx,#
100/   ldy,#
$ffbd  jsr, \ setname
$f     lda,#
2      ldx,#
0      ldy,#
$ffba  jsr, \ setlfs
$ffc0  jsr, \ open
$f     ldx,#
$ffc6  jsr, \ chkin
$ffe4  jsr, \ getin toss null
$ffcc  jsr, \ clrchn 
8      lda,#
$ffd2  jsr, '1' @: \ lower case
$ffcc  jsr, \ clrchn
$ffe4  jsr, \ getin
0      cmp,# +branch
       beq, \ keyboard buffer empty?
       pha, \ no
$ffcc  jsr, \ clrchn
$f     ldx,#
$ffc9  jsr, \ chkout
       pla,
$ffd2  jsr, :+
$ffcc  jsr, \ clrchn
$f     ldx,#
$ffc6  jsr, \ chkin
$ffe4  jsr, \ getin  loop
0      cmp,# '1' @@
       beq, \ receive buffer empty?
$ffd2  jsr, here $0a - \ no
       jmp,  \ loop
w      ldx,
;code
```
![vice-screen-20220110014008](https://user-images.githubusercontent.com/38296423/148749755-c6bfed56-982c-49ec-a7a3-fd305adbb2a0.jpg)

**Comment by Whammo** [comment]
Settings for Vice. Try  bbs.retroacademy.it:6510

![image](https://user-images.githubusercontent.com/38296423/148772901-48d4c205-fe56-4940-a8c1-48148ff920b8.png)

**Comment by Whammo** [comment]
Implications:
ASCII translation
Download protocol 
REFILL from device 2
REMOTE control
~~Word for GETIN~~
Ideas?

**Reply by Whammo** [reply]
It actually compiles- I just left some spaces out when it was documented.

**Comment by Whammo** [comment]
~~Now to determine a method for a graceful exit. This may be problematic~~ in Vice as it doesn't seem to want to close the connection, or at least allow one to open it again after closing.

**Comment by Whammo** [comment]
For serial GETIN, the accumulator returns the byte received over the serial bus. However, if any I/O status errors occur, return with accumulator containing $0D. So a CR is also an event that should be checked with ST.

**Comment by Whammo** [comment]
Both forth and asm versions drop/corrupt characters at 2400 baud. Not happening at 1200.
 George Hug routines seem like overkill. 
REFILL from rs232 at 2400 will likely be good with CR - one line at a time- handshaking

**Comment by Whammo** [comment]
Something that might be useful for an actual terminal program.
 ```forth
code bell
$15    lda,#
$d418  sta,
$09    ldy,#
0      ldx,#
$d405  sty,
$d406  stx,
$30    lda,#
$d401  sta,
$20    lda,#
$d404  sta,
$21    lda,#
$d404  sta,
;code
```

**Comment by Whammo** [comment]
~~DT in Forth has ST race condition, closes connection before emptying receive buffer.~~

**Comment by Whammo** [comment]
Started experiments with TCPserial. Probably will end up starting again from scratch.

**Comment by cminter** [comment]
@Whammo,
Thanks for sharing.  **Really nice**.  I need to study this.  I've wanted to try to do a _customized_ terminal _in Forth_ ever since I rediscovered my WiFi modem that I had stuck in a box.

**Reply by Whammo** [reply]
Thanks.
There is quite a bit of history and some inside technical information pertaining to rs232 on durexForth deep in discussions.

**Reply by cminter** [reply]
Yea I've noticed you have a lot of code gems embedded everywhere.

**Comment by Whammo** [comment]
An important but unstated goal the option to have durexForth headless, I/O redirected through the rs232 port.

**Comment by jkotlinski** [comment]
Cool idea!

fre 31 jan. 2025 kl. 14:04 skrev Whammo ***@***.***>:
